src/main/java/com/linecorp/armeria/server/grpc/ArmeriaGrpcServerStream.java

```diff
@@ -0,0 +1,140 @@
+package com.linecorp.armeria.server.grpc;
+
+import java.nio.charset.StandardCharsets;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.http.DefaultHttpHeaders;
+import com.linecorp.armeria.common.http.HttpData;
+import com.linecorp.armeria.common.http.HttpHeaderNames;
+import com.linecorp.armeria.common.http.HttpHeaders;
+import com.linecorp.armeria.common.http.HttpResponseWriter;
+import com.linecorp.armeria.common.http.HttpStatus;
+import com.linecorp.armeria.internal.grpc.NettyWritableBuffer;
+import com.linecorp.armeria.internal.grpc.NettyWritableBufferAllocator;
+
+import io.grpc.Metadata;
+import io.grpc.Status;
+import io.grpc.internal.AbstractServerStream;
+import io.grpc.internal.GrpcUtil;
+import io.grpc.internal.TransportFrameUtil;
+import io.grpc.internal.WritableBuffer;
+import io.netty.buffer.ByteBuf;
+import io.netty.util.AsciiString;
+
+/**
+ * An {@link AbstractServerStream} that converts GRPC structures and writes to an Armeria
+ * {@link HttpResponseWriter}. GRPC will have already taken care of compression, framing, etc.
+ */
+class ArmeriaGrpcServerStream extends AbstractServerStream {
+
+    private final ArmeriaMessageReader messageReader;
+    private final HttpResponseWriter responseWriter;
+    private final Sink sink;
+    private final TransportState transportState;
+    private final long maxMessageSize;
+
+    ArmeriaGrpcServerStream(HttpResponseWriter responseWriter,
+                                      long maxMessageSize) {
+        super(new NettyWritableBufferAllocator());
+        this.responseWriter = responseWriter;
+        this.maxMessageSize = maxMessageSize;
+        this.sink = new Sink();
+        this.transportState = new TransportState();
+        this.messageReader = new ArmeriaMessageReader(transportState);
+    }
+
+    ArmeriaMessageReader messageReader() {
+        return this.messageReader;
+    }
+
+    @Override
+    protected TransportState transportState() {
+        return transportState;
+    }
+
+    @Override
+    protected Sink abstractServerStreamSink() {
+        return sink;
+    }
+
+    private class Sink implements AbstractServerStream.Sink {
+
+        @Override
+        public void writeHeaders(Metadata headers) {
+            responseWriter.write(convertMetadata(headers));
+        }
+
+        // Armeria always flushes so we ignore the flush parameter.
+        @Override
+        public void writeFrame(@Nullable WritableBuffer frame, boolean flush) {
+            if (frame == null) {
+                // Armeria flushes every message so no need to ever manually flush.
```

Comment:
This comment was a replacement for something like writer.flush() so I put it here. Extended it to see if it's clearer.